### ISSUE TITLE
publication version changes order of traces in trace list

### ISSUE THREAD

**Issue by krischer** [iris-edu/libmseed#44]
Files that have records with different publication versions and are being read with `mstl3_readbuffer_selection()` will result in traces that are ordered front to back. This is not a deal breaker but I feel this should not be the case.

This file (https://www.dropbox.com/s/sibg35lgb4t7t0x/test.ms3?dl=1) contains three records from stations `AA`, `BB`, and `CC`. publications versions go from 1 to 3.

The resulting trace list is in the correct order if the publication versions are identical but as soon as they differ the order is reversed.

**Comment by chad-iris** [comment]
I think I see what you're saying.  More generally, if the traces are being split on the publication number, then the ordering of the traces appears to be dependent on the order they were added to the trace list.    This breaks a core tenant of the library, which is that order of input should not matter at all, and order of presentation should be consistent.

**Comment by krischer** [comment]
Not exactly what I was trying to express but I see your point and it makes more sense than mine. It would be nice if the ordering would be predictable in the sense that its first sorted alphabetically by SID, then by publication version, and then by segment start time or something like it.

Out of curiosity: In what scenario would one ever not split on the publication version? And would this then result in a single `MS3TraceID` having multiple `MS3TraceSeg` with different publication versions? Is the `pubversion` field on the trace id then invalid?

**Comment by chad-iris** [comment]
> Out of curiosity: In what scenario would one ever not split on the publication version?

I recommend to _not_ split on publication version unless you have some special reason to separate them.  I will add that recommendation to the docs.

The very likely scenario: an network operator replaces a subset of data with a later version and increases the publication version number of the replacement data.  A data center sends a user data for a selection that includes both the old and new data so the user gets a mix of versions, which are otherwise a continuous time series.  This is what happened for years with quality identifiers.  You don't really want to split them as that will just make more work for the user to put them back together for processing.

> And would this then result in a single MS3TraceID having multiple MS3TraceSeg with different publication versions?

When splitting, each new publication version creates a new MS3TraceID.  Currently, that means there can be multiple MS3TraceIDs with the same SID value, and this is why they didn't get sorted.

I'm thinking I'll add **"#VERSION"** to the SID when splitting on publication version.  This is equivalent to what version 2 did when splitting on data quality.

> Is the pubversion field on the trace id then invalid?

As documented in mstl3_addmsr():
"If more than one version contributes to a given source identifer's segments, its publication version will be the set to the largest contributing version."

**Comment by chad-iris** [comment]
Addressed in 78e6d8e8bab69a7ea569b574a04fbec96091d241.

I decided not to annotate, and thus pollute, the SIDs.  Also, it would have been redundant as the MS3TraceID already has the publication version.

Current behavior: if splitting publication versions, each version is isolated to separate MS3TraceID entries and adjacent MS3TraceIDs that have the same SID are maintained in ascending publication version order.

I also documented the recommended behavior of _not_ splitting on publication versions.  A main point of this routine is to construct the best, most continuous traces possible.

One could argue that splitting on publication version should not even be allowed, I leave it in for those of us that might need to do something in the future like comparing versions of the "same" data.
